dali/operators/expressions/expression_impl_factory.h

```diff
@@ -21,80 +21,150 @@
+ */
+template <typename Backend>
+inline ArgPack GetArgPack(const ExprFunc &func, workspace_t<Backend> &ws,
+                          const ConstantStorage<Backend> &st, const OpSpec &spec, TileDesc tile) {
+  ArgPack result;
+  result.resize(func.GetSubexpressionCount());
+  for (int i = 0; i < func.GetSubexpressionCount(); i++) {
+    DALI_ENFORCE(func[i].GetNodeType() != NodeType::Function,
+                 "Function nodes are not supported as subexpressions");
+    if (func[i].GetNodeType() == NodeType::Constant) {
+      const auto &constant = dynamic_cast<const ExprConstant &>(func[i]);
+      result[i] = st.GetPointer(constant.GetConstIndex(), constant.GetTypeId());
+    } else if (func[i].GetNodeType() == NodeType::Tensor) {
+      const auto &tensor = dynamic_cast<const ExprTensor &>(func[i]);
+      auto input_idx = tensor.GetInputIndex();
+      const auto *ptr =
+          reinterpret_cast<const char *>(GetInputSamplePointer(ws, input_idx, tile.sample_idx));
+      auto tile_offset =
+          tile.tile_size * tile.extent_idx * TypeTable::GetTypeInfo(func.GetTypeId()).size();
+      result[i] = ptr + tile_offset;
+    }
+  }
+  return result;
+}
+
+/**
+ * @brief Transfor vector of TileDesc into vector of ExtendedTileDesc
+ * based on the ExprFunc by extracting the input and output pointers to data
+ * from workspace and constant storage.
+ *
+ * @param extended_tiles Output vector of ExtendedTiles for given task
+ */
+template <typename Backend>
+void TransformDescs(std::vector<ExtendedTileDesc> &extended_tiles,
+                           const std::vector<TileDesc> &tiles, const ExprFunc &func,
+                           workspace_t<Backend> &ws, const ConstantStorage<Backend> &st,
+                           const OpSpec &spec) {
+  extended_tiles.reserve(tiles.size());
+  for (auto &tile : tiles) {
+    extended_tiles.emplace_back(tile, GetOutput<Backend>(func, ws, tile),
+                                GetArgPack(func, ws, st, spec, tile));
+  }
+}
+
+/**
+ * @brief Prepare vector of ExtendedTiles for every task that we have to execute, filling
+ * the pointers to data.
+ *
+ * @param tiles_per_task  Output vectors of ExtendedTiles per evry tast to execute
+ */
+template <typename Backend>
+void PrepareTilesForTasks(std::vector<std::vector<ExtendedTileDesc>> &tiles_per_task,
+                          const std::vector<ExprImplTask> &task_exec_order,
+                          const std::vector<TileDesc> &tiles, workspace_t<Backend> &ws,
+                          const ConstantStorage<Backend> &constant_storage, const OpSpec &spec) {
+  tiles_per_task.resize(task_exec_order.size());
+  for (size_t i = 0; i < task_exec_order.size(); i++) {
+    const auto &expr_task = task_exec_order[i];
+    const auto &expr_func = dynamic_cast<const ExprFunc &>(*expr_task.ctx.node);
+    tiles_per_task[i].resize(0);
+    TransformDescs<Backend>(tiles_per_task[i], tiles, expr_func, ws, constant_storage, spec);
```

Comment:
If I understand correctly for every element of `task_exec_order` we go over all `tiles`?

Reply:
Yes.
We have some data to process, we tile based on output shape, and to produce each tile of output we need to process all tasks. Still, the number of task is currently equal to 1.